utils/property-propagation.stanza

```diff
@@ -0,0 +1,220 @@
+          match(get?(nets-by-pin, pin)) : 
+            (f:False) : 
+              ; println("%_ is unconnected." % [ref(pin)])
+              ; the pin is unconnected, do nothing.
+              false 
+            (n:JITXObject) : 
+              ; println("%_ is connected to %_, which is visited?: %_" % [ref(pin), n, visited?[n]])
+              ; the pin is connected to a net we haven't visited yet.
+              add(net-stack, n) when not visited?[n]
+
+; same as above with default updater
+public defn propagate-property-along-nets (root-module:JITXObject,
+                                           property-name:Symbol,
+                                           handle-instance: (JITXObject) -> Seqable<JITXObject>,
+                                           handle-conflict:(Seqable<jitx/client/JITXValue>) -> jitx/client/JITXValue) :
+  propagate-property-along-nets(root-module, property-name, handle-instance, handle-conflict, default-updater)
+
+; same as above with default handlers
+public defn propagate-property-along-nets (root-module:JITXObject,
+                                           property-name:Symbol) :
+  propagate-property-along-nets(root-module, 
+                                property-name, 
+                                default-instance-handler, 
+                                default-conflict-handler)
+
+; WIP: new rail-voltage propagation based on a two-pass approach
+; 
+; Example: 
+;   Here we propagate the "power-supply-pin" property along nets
+;   connected to any pin marked with this property, but map the 
+;   maximum voltage of the psu-pin to a property called "rail-voltage."
+; 
+;   All we need to do is call the main driver and define our handlers, 
+;   in this case we can use the default-instance-handler (which does 
+;   not propagate across components), with a custom conflict handler 
+;   and custom updater.
+;
+public defn propagate-rail-voltage-2 (root-module:JITXObject) : 
+  propagate-property-along-nets(root-module,
+                                `power-supply-pin,
+                                default-instance-handler,
+                                handle-conflict,
+                                updater) where :
+    defn handle-conflict (xs) : 
+      maximum $ 
+        for psp in filter-by<PowerSupplyPin>(xs) seq : 
+          match(supply-voltage(psp)) : 
+            (d:Double)     : d
+            (t:Toleranced) : max-value(t)
+
+    defn updater (p:JITXObject, name:Symbol, value:jitx/client/JITXValue) : 
+      set-property(p, `rail-voltage, value)
+    
+;==============================================================================
+;============================= Helpers ========================================
+;==============================================================================
+; TODO: refactor not to use undocumented internal APIs
+
+doc:"Returns if an object has a property."
+defn has-property? (obj:JITXObject, prop:Symbol) -> True|False : 
```

Comment:
Add this function to internal API instead of hacking it through undocumented APIs

Reply:
This is undocumented? Wow. As it is used liberally in checks, for example.

Reply:
has-property?(thing.property) is what we use extensively, but has-property?(thing,`property) as defined here uses internals that aren't exposed normally

Reply:
Ok